### ISSUE TITLE
Windows 10 mouse scroll issue

### ISSUE THREAD

**Issue by dcnieho** [bvschaik/julius#172]
When trying to scroll rightward (or downward), it does not work (view doesn't change) when i place my cursor as far right (or down) as it can go. If i put it one pixel inward, scrolling happens. No problem with the left and upper screen edges.

I don't seem to have a log file to send you?

This is on fully updated Windows 10 Enterprise, using the nightly build from yesterday. Happen both on my laptop's built in screen, and on an external screen (both single monitor setups, so internal screen disabled when on external screen).

Let me also say that it is amazing to be able to play the game on my big monitor, super cool! How did you manage to achieve this, if i may ask? Partial decompile and make your own new graphics and sound back end? I guess there is large overlap in engine between these and the Pharaoh, Zeus and ROTMK games, so perhaps in some distant future we may get those to work too. I am not in a position to offer help with that now, but perhaps. Thanks for bringing this back to me, its impressive, really impressive how well it all works!

**Comment by bvschaik** [comment]
Are you playing full screen or windowed? Since you're using display scaling, Windows might not consider the very last pixel as part of the game's window due to rounding. I'll have a look.
The log file `julius-log.txt` _should_ be in the same directory as your julius.exe (which works best if it's in the same directory as your C3 installation). If it's not there I will have a look at that as well.

**Comment by crudelios** [comment]
@bvschaik long time no see! I've been away for a while and slowing inching to get back to julius, but the amount of work I have isn't helping...

Anyway, both this issue and #173 are due to DPI scaling.

I had both issues fixed in my DPI PR, but that, in turn, created many other issues so I removed the PR.

Still, the only solution I can think of is to take care of DPI manually, because SDL doesn't do that (at least it didn't in v2.0.9).

**Comment by rsn8887** [comment]
Couldn't you just check for last or second-to last pixel, e.g. use a two-pixel active boundary?

**Comment by dcnieho** [comment]
If you point me to the relevant bit in the code for windows, i'm happy to have a look and try out some possible fixes if i see any.

@bvschaik: I play fullscreen (so no window decorations) everything else is normal size, just the cursor is so small its hard to see.
@crudelios: in that sense, i guess changes to DPI awareness aren't necessary other than for the cursors? Or is there other things i might notice if i look carefully?

regarding log: I guess it is a permissions issue. Admin rights (or a UAC accept) are needed to create files in the program files folder. If i run the exe from my desktop, a log appears. here's its contents:

```
INFO: Julius version 1.1.0.129-cec9a3fb
INFO: Initializing SDL
INFO: SDL initialized
INFO: Loading game from working directory
ERROR: 'c3.eng' or 'c3_mm.eng' files not found or too large.
INFO: Loading game from base path C:\Users\huml-dkn\Desktop\
ERROR: 'c3.eng' or 'c3_mm.eng' files not found or too large.
INFO: Loading game from user-selected dir C:\Program Files (x86)\Impressions Games\Caesar3
INFO: Detected encoding:  1252
INFO: Creating screen 1536 x 864, fullscreen? 1
INFO: Texture created (1536 x 864)
INFO: model loaded
INFO: Window resized to 1536 x 864
INFO: Texture created (1536 x 864)
INFO: System resize to 1536 x 864
INFO: Window 1 shown
INFO: Loading saved game  Architect6.sav
INFO: Exiting game
```

**Comment by bvschaik** [comment]
@crudelios Long time no see indeed! Windows appears to handle the scaling when you're not enabling HIGHDPI, *except* for scaling the cursor, and the edge problem when scrolling. We can perhaps fix the cursor by scaling it manually. Did you ever find out why the screen edges are not recognised by SDL? Does it not pass mouse events for the border or is it something different?
And what were the results on Linux? Similar to Windows or does it have its own problems?
On Mac there are no issues at all on Retina displays. Ideally I'd like the game to be DPI-unaware, since your attempt to make it DPI-aware created more issues than it solved...

@dcnieho The scrolling code is in `src/input/scroll.c`. I suggest putting some logging in there to see why it doesn't recognise the screen edges.
Regarding the log: yes, if your C3 install is in the program files the log will not be written. But on the other hand: do saved games get saved properly?

**Comment by dcnieho** [comment]
ah, my save games (and log) appear
in C:\Users\huml-dkn\AppData\Local\VirtualStore\Program Files
(x86)\Impressions Games\Caesar3 (and there is nothing else in the
VirtualStore tree interestingly enough, and i've never heard of this
before--which may not mean much :p). So yeah, it all works fine.
I'll try and have a look at the scroll this week.

**Comment by crudelios** [comment]
@dcnieho The cursor size and the edges were the only issues I noticed.

@bvschaik The only way I found to get the current system DPI and figure out the required scaling value was to make tge app DPI-aware, otherwise Windows/SDL always throw out the default value of 96/1.0, simulating no sc scaling. However making the app DPI-aware forces you to handle DPI changes manually, so that's a catch-22!
The reason the game doesn't detect scrolling is because it considers the bottommost and rightmost edges of the screen to be the borders of the window, so SDL doesn't even register the mouse event at all. I never understood why making the app DPI-aware fixed the issue, but it did!

In conclusion, both bugs are caused by SDL.

**Comment by dcnieho** [comment]
Ok, added some logging to `handle_window_event()` and `scroll_get_direction()`. Output on my fullscreen window is:

```
INFO: Creating screen 1536 x 864, fullscreen? 1
INFO: Texture created (1536 x 864)
INFO: model loaded
INFO: Window resized to 1536 x 864
INFO: Texture created (1536 x 864)
INFO: System resize to 1536 x 864
INFO: Window 1 shown

INFO: mouse inside? 1. position: 746 x 858, screen: 1536 x 864
INFO: mouse inside? 1. position: 746 x 860, screen: 1536 x 864
INFO: mouse inside? 1. position: 746 x 862, screen: 1536 x 864
INFO: mouse inside? 1. position: 746 x 862, screen: 1536 x 864
INFO: got SDL_WINDOWEVENT_LEAVE
INFO: mouse inside? 0. position: 746 x 863, screen: 1536 x 864
INFO: mouse inside? 0. position: 746 x 863, screen: 1536 x 864
INFO: mouse inside? 0. position: 746 x 863, screen: 1536 x 864
INFO: got SDL_WINDOWEVENT_ENTER
INFO: mouse inside? 1. position: 746 x 863, screen: 1536 x 864
INFO: mouse inside? 1. position: 746 x 861, screen: 1536 x 864
INFO: mouse inside? 1. position: 746 x 858, screen: 1536 x 864
```

So i added the following hack atop `scroll_get_direction()`:
```
    int is_inside_window = m->is_inside_window;
    int width = screen_width();     // moved from lower in the function
    int height = screen_height();
    if (setting_fullscreen() && m->x < width && m->y < height)
    {
        is_inside_window = 1;
    }
```
And then use is_inside_window  instead of m->is_inside_window  in the rest of the function. Everything works well for me that way. I do not know how this would play on other platforms or with multi-monitor (latter i can test if you like).

Is this a decent workaround (should i sent a PR?), or likely to cause havoc?
From what little i know about the issue, if enabling DPI awareness leads to trouble (and except for small cursor, everything works great now on Windows even at my 250% scale!), perhaps a workaround or two like this is what we need? And perhaps confine it to Windows?

I used the latest SDL (2.0.10) by the way.

**Comment by crudelios** [comment]
I've looked at SDL today and discovered this is a issue with Windows, specificaly with the `CreateIconIndirect` function. AFAIK there is no fix for it.

A simple, alternative workaround it to allow the user to set the mouse cursor scaling, for example via command line.

**Comment by dcnieho** [comment]
Hi @crudelios :) Thats exactly something i just pushed (and am still fixing up, the CI builds are catching a bunch of things my build system is not)

**Comment by bvschaik** [comment]
@dcnieho The mouse scroll issue is fixed now, right?

**Comment by dcnieho** [comment]
To be sure, i just checked out master and tested. looking good!
